utility/RTC_Super.h

```diff
@@ -0,0 +1,74 @@
+/**************************************************************************/
+/*!
+  @file     RTC_Super.h
+
+  Original library by JeeLabs http://news.jeelabs.org/code/, released to the
+  public domain
+
+  License: MIT (see LICENSE)
+
+  This is a fork of JeeLab's fantastic real time clock library for Arduino.
+
+  For details on using this library with an RTC module like the DS1307, PCF8523,
+  or DS3231, see the guide at:
+  https://learn.adafruit.com/ds1307-real-time-clock-breakout-board-kit/overview
+
+  Adafruit invests time and resources providing this open source code,
+  please support Adafruit and open-source hardware by purchasing
+  products from Adafruit!
+*/
+/**************************************************************************/
+
+#ifndef _RTC_SUPER_H_
+#define _RTC_SUPER_H_
+
+#include "DateTime.h"
+#include <Arduino.h>
+
+/**************************************************************************/
+/*!
+    @brief  RTC superclass for all RTC chips
+*/
+/**************************************************************************/
+class RTC_Super {
```

Comment:
I originally called it `RTC` but there's already a macro of the same name. Perhaps `RTC_Base`?

Reply:
Where is that `RTC` macro? I couldn't find it.

Reply:
I would just change the macro. It really should be called RTC.

Reply:
When I left it as `RTC` the automated test script detected an error with the `RTC` macro in
/home/runner/.arduino15/packages/arduino/tools/CMSIS-Atmel/1.2.0/CMSIS/Device/ATMEL/samd51/include/samd51j19a.h:957
You can see it in the Checks tab.

Reply:
Is Arduino.h needed?

Here is the file with the macros:
https://github.com/adafruit/ArduinoModule-CMSIS-Atmel/blob/master/CMSIS-Atmel/CMSIS/Device/ATMEL/samd51/include/samd51j19a.h

Here is possibly were the get included from:
https://github.com/arduino/ArduinoCore-samd/blob/master/cores/arduino/Arduino.h

https://github.com/adafruit/RTClib/blob/e2a981c7d9115b8e53ac62581e01ffd755c6e073/utility/RTC_Super.h#L26

Reply:
You could try using extern "C".

Reply:
@daveythacher I haven't looked into it rigorously, but I don't think the RTC base class needs `Arduino.h`. However, the derived RTC classes do need `Arduino.h` so the errors between a class and a macro both being called `RTC` will persist in `RTClib.h`.

Reply:
Look into the scope of macros in the preprocessor. https://stackoverflow.com/questions/6379489/scope-of-define-preprocessor-in-c

This is a good discussion: https://stackoverflow.com/questions/2321713/how-do-i-avoid-name-collision-with-macros-defined-in-windows-header-files

C++ does not get along with macros, but I still say you can work around this. It will not be clean, but should be possible. There are a few cases here worth mentioning. 

Some devices have RTC modules which will not be used and conflict with the library. For these we can probably just undef the macro globally.

Some devices have RTC modules which will be used and conflict with the library. For these you would have c functions prototypes defined. Include the RTC.h, then setup the C++ method logic which calls c functions. Below that include Arduino,h and implement the c functions.

If I am not mistaken that will remove the conflict. There may be better ways. However the main thing here is the macro lives from instance to end of file. So if you included it at different point it will not exist before that point to my understanding. You can also terminate it early with undef. There is a few other approaches mentioned in that discussion.

Reply:
This would also be a tad easier if those drivers were broken out to there own file.

Reply:
Given that inheritance in C++ is meant to represent an “is a” relationship, what if we said that every concrete RTC “is a `RealTimeClock`”? It sound better to my ears than “is an `RTC_Super`”.

Reply:
@edgar-bonet what about putting this code at the start/end of the `RTC.h` and `RTClib.h` files?
```c++
#pragma push_macro("RTC")
#undef RTC
...
#pragma pop_macro("RTC")
```

I'm stuck between the above and using `RealTimeClock` as the base class name.

Reply:
Actually, nevermind. Using `#pragma` is not beginner friendly.